Add 'Send to Kindle' feature

aqurilla opened the pull request:
<!-- 
Describe the changes you have made here: what, why, ... 
Link issues that are fixed, e.g. "Fixes #333".
If you fixed a koppor issue, link it, e.g. "Fixes https://github.com/koppor/jabref/issues/47".
The title of the PR must not reference an issue, because GitHub does not support autolinking there.
-->

This fixes #6186 by adding 'Send to Kindle' context & Tools menu actions. The user's Kindle email can be saved in preferences under the 'External Programs' section.

<img width="1101" alt="image" src="https://github.com/JabRef/jabref/assets/12078268/f228a53a-17be-45b4-b406-b5ba3169eecd">
<img width="724" alt="image" src="https://github.com/JabRef/jabref/assets/12078268/5bb4e8d3-fc46-4e06-b57c-f1d503bbb256">
<img width="835" alt="image" src="https://github.com/JabRef/jabref/assets/12078268/f8a1db09-5963-426a-80b4-361ca8e63c78">

<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

### Mandatory checks
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

Siedlerchr reviewed (dismissed):
Codewise looks good so far!  Haven't tested it yet.

koppor commented on src/main/java/org/jabref/gui/SendAsEMailAction.java:
```diff
@@ -107,19 +111,30 @@ private String sendEmail() throws Exception {
+
+    private URI getUriMailTo(StringWriter rawEntries, List<String> attachments) throws URISyntaxException {
+        StringBuilder mailTo = new StringBuilder();
+        if (action == StandardActions.SEND_TO_KINDLE) {
+            mailTo.append(preferencesService.getExternalApplicationsPreferences().getKindleEmail());
+            mailTo.append("?Subject=").append(Localization.lang("Send to Kindle"));
+        } else if (action == StandardActions.SEND_AS_EMAIL) {
+            mailTo.append("?Body=").append(rawEntries.toString());
+            mailTo.append("&Subject=");
+            mailTo.append(preferencesService.getExternalApplicationsPreferences().getEmailSubject());
+        }
```
Just a comment:

This is not object-oriented. Normally, one would crreate an abstract class and two sub classes, where the subclasses differ in that behavior.

However, these are 5 lines of Code only, so OK for me.

koppor commented on src/main/java/org/jabref/gui/SendAsEMailAction.java:
```diff
@@ -107,19 +111,30 @@ private String sendEmail() throws Exception {
+        StringBuilder mailTo = new StringBuilder();
+        if (action == StandardActions.SEND_TO_KINDLE) {
+            mailTo.append(preferencesService.getExternalApplicationsPreferences().getKindleEmail());
+            mailTo.append("?Subject=").append(Localization.lang("Send to Kindle"));
```
I would also add "Send to kindle" as email body. Otherwise, an empty email could be rejected. Who knows...

aqurilla replied:
Sure, will add that. I did try it out using my Kindle email and it does not require a subject or body, but agreed that it would be best to future proof ourselves by adding both anyways 😄

koppor replied:
Checking for attachments is enough. Enforced Email configuration could be too hard for newcomers to JabRef.

koppor commented on src/main/java/org/jabref/gui/SendAsEMailAction.java:
```diff
@@ -107,19 +111,30 @@ private String sendEmail() throws Exception {
+    private URI getUriMailTo(StringWriter rawEntries, List<String> attachments) throws URISyntaxException {
+        StringBuilder mailTo = new StringBuilder();
+        if (action == StandardActions.SEND_TO_KINDLE) {
+            mailTo.append(preferencesService.getExternalApplicationsPreferences().getKindleEmail());
```
I assume this also works if the email is not configured?

aqurilla replied:
Currently it does work without a configured email and opens the email dialog with a blank mailTo address. Should this behavior be changed?

Siedlerchr replied:
Hm, I think it's oaky. one can still add the mail in the email program

koppor reviewed (changes requested):
In general OK. (And on my machine Send as email stopped working in general, no issue of this PR)

The functionality should only be enabled if there is an attachment:

![image](https://github.com/JabRef/jabref/assets/1366654/0cdfeceb-efc7-45c8-8e7a-e99597c8f370)

The whole point is that the file is send ^^. Thus, please add an enablement check of this menu item (in case attachemnt are present, it is enabled. Otherwise, not)

Siedlerchr commented:
Suggestion: Integrate both "Mail" actions into a new submenu Send. 
e.g. Send.... 
         + mail
         + to kindle...

aqurilla commented:
Added a Send submenu with the email and Kindle options, and added the file attachment check for the Kindle option.
Also converted the different email actions to an abstract class and subclasses as per the recommendation.

<img width="698" alt="image" src="https://github.com/JabRef/jabref/assets/12078268/dd8d01c5-18fd-482b-87ea-4d827f3b90ea">

koppor commented on src/main/java/org/jabref/gui/SendAsKindleEmailAction.java:
```diff
@@ -0,0 +1,39 @@
+    @Override
+    protected URI getUriMailTo(StringWriter rawEntries, List<String> attachments) throws URISyntaxException {
+        StringBuilder mailTo = new StringBuilder();
+        mailTo.append(preferencesService.getExternalApplicationsPreferences().getKindleEmail());
+        mailTo.append("?Body=").append(Localization.lang("Send to Kindle"));
+        mailTo.append("&Subject=").append(Localization.lang("Send to Kindle"));
```
These are the lines differing at the sub classes.

Proposal: Move up `getUriMailTo` and create three new abstract methods:

- `String getEmailAdress()`
- `String getSubject()`
- `String getBody()`

In this way, each sub class takes care about its differentiating thing - and the abstract class summarizes the things not being differen.t

aqurilla replied:
Got it thanks, I've updated the PR with these changes

koppor commented on src/main/java/org/jabref/gui/SendAsGeneralEmailAction.java:
```diff
@@ -0,0 +1,38 @@
+ * Sends the selected entries to any specifiable email
+ * by populating the email body
+ */
+public class SendAsGeneralEmailAction extends SendAsEMailAction {
```
I don't like `General`. Maybe `Standard` is better? --> `SendAsStandardEmailAction`

koppor commented on src/main/java/org/jabref/gui/preferences/external/ExternalTab.fxml:
```diff
@@ -21,6 +21,10 @@
         <TextField fx:id="eMailReferenceSubject" minWidth="150.0" HBox.hgrow="ALWAYS"/>
     </HBox>
+    <HBox alignment="CENTER_LEFT" spacing="10.0">
+        <Label text="%Send-to-kindle email address"/>
```
Maybe `Email address for sending to Kindle`? In this way, the term what to input comes first ("Email address") and then the concretization ("sending to Kindle")

koppor commented on src/main/java/org/jabref/gui/SendAsEMailAction.java:
```diff
@@ -107,19 +105,32 @@ private String sendEmail() throws Exception {
+
+    protected abstract String getSubject();
+
+    protected abstract String getBody(StringWriter rawEntries);
```
I think, the decision is between performance or code readability.

First suggestion - then the stringwriter is converted at the caller

```suggestion
    protected abstract String getBody(String rawEntries);
```

Second suggestion: As only one sub class needs the bibtex entries, I would pass the List of bibtex entries. The standard email action can do the conversion (you have to move the code from lines 74 to 89) to there. And the kindle action can ignore it.

aqurilla replied:
Sounds good. For the second suggestion we would need both the entries and databaseContext in `getBody` - is it preferable to get those from `stateManager` in the subclass instead of passing them in as arguments?

calixtus replied:
The stateManager can deliver the currently active database and also the currently selected Entries.

Could be a problem though if you right-click on any other entry and you want to send the clicked one. Windows explorer does unselect in this case all the other entries and only the one right-clicked on. So this should be none of your concern in this PR. I think `stateManager.getActiveDatabase()` and `stateManager.getSelectedEntries()` should be sufficient.

aqurilla replied:
Thanks for clarifying that! I'll go ahead with this approach

koppor reviewed (changes requested):
Thank you for introducing the class hierarchy. One more thing can be moved.

koppor reviewed (changes requested):
Thank you for keeping working on this. One nitpick comment ^^. Otherwise: Looks-good-to-me!

aqurilla commented:
Thanks, appreciate it!
